keystore/v2/keystore/api/key.go

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2020, Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/cossacklabs/acra/keystore/v2/keystore/asn1"
+)
+
+// Errors returned by Key methods:
+var (
+	ErrFormatDuplicated    = errors.New("KeyStore: key format used multiple times")
+	ErrFormatMissing       = errors.New("KeyStore: key format not available")
+	ErrKeyNotExist         = errors.New("KeyStore: no key with such seqnum")
+	ErrNoKeyData           = errors.New("KeyStore: no key data")
+	ErrInvalidFormat       = errors.New("KeyStore: invalid key format")
+	ErrInvalidState        = errors.New("KeyStore: invalid state transition")
+	ErrInvalidCryptoperiod = errors.New("KeyStore: invalid key cryptoperiod")
+)
+
+// Key in a KeyRing.
+type Key interface {
+	// Seqnum returns sequential number of this key in the key ring.
+	Seqnum() (int, error)
+	// State of the key right now.
+	State() (KeyState, error)
+
+	// ValidSince returns the time before which the key cannot be used.
+	ValidSince() (time.Time, error)
+	// ValidUntil returns the time since which the key should not be used.
+	ValidUntil() (time.Time, error)
+
+	// Formats available for this key.
+	Formats() ([]KeyFormat, error)
+	// PublicKey data in given format, if available.
+	PublicKey(format KeyFormat) ([]byte, error)
+	// PrivateKey data in given format, if available.
+	PrivateKey(format KeyFormat) ([]byte, error)
+	// SymmetricKey data in given format, if available.
+	SymmetricKey(format KeyFormat) ([]byte, error)
+}
+
+// MutableKey in a KeyRing.
+type MutableKey interface {
```

Comment:
looked one more time and can't imagine situation when key's state should be changed by itself. always it is the responsibility of keyring. when new key added, keyring take new key and set state pre-active or active (depends on ValidSince value. If key is compromised, our some cli tool will accept some key that should be marked as compromised and passed to keyring/keystore to mark. Keyring set new state (and probably place key at end of keychain). If keystore process GetKey request and see that new key from keychain is outdated, it set status Deactivate, place at end of keychain and return another valid key.

Plus looked on our key statuses. I think that good to have only statuses that can be set manually after some event.
We don't need pre-active/active/deactivated statuses because they depends on time period of key (valid since, valid 
 until) and should be configured automatically by keyring/keystorage.

Reply:
> looked one more time and can't imagine situation when key's state should be changed by itself. always it is the responsibility of keyring

Alrighty then. I’ll clean up and add to this PR that change which moves all key access API into the key ring directly.

> We don't need pre-active/active/deactivated statuses because they depends on time period of key (valid since, valid until) and should be configured automatically by keyring/keystorage.

Hm... I don’t think that validity range coincides with usage timespan. The key should not be used outside of validity range (either before or after it), but it can be active for a shorter period, as defined by some usage policy.

For example, some keys may be generated in advance and placed into _pre-active_ state, with validity range starting on 2020-04-01 and ending 2021-03-31. However, something comes up in the infrastructure and it is not possible to start using keys on 2020-04-01. Because of ${legal_reasons}, IDK. Similarly, it might be necessary to deactivate a key earlier that its scheduled validity end because a new policy says that keys should not be valid for more than _half_ a year, not a whole year as before. These cases call for explicit key activation and deactivation by administrator, not implicit activation by the key store itself.

The key store provide a mechanism (tracking key states), it’s up to users to decide on the policy of the state updates.

(That being said, it’s probably reasonable to deactivate a key or otherwise complicate its usage past the validity range. Though again I wonder whether this should be enforced by elsewhere or by the key store directly.)

Reply:
Okay, status processing and managing we will see when will integrate it into acra. for now more generalized better than more limited